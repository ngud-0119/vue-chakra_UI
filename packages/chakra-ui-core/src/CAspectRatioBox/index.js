import AspectRatioBox from './AspectRatioBox'
export default AspectRatioBox
