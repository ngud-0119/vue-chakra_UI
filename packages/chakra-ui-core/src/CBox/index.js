import CBox from './CBox'
export default CBox
export * from './CBox'
