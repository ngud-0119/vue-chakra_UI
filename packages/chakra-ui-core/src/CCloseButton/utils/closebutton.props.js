export default {
  size: {
    type: String,
    default: 'md',
    validator: value => value.match(/^(sm|md|lg)$/)
  },
  isDisabled: {
    type: Boolean,
    default: false
  },
  ariaLabel: {
    type: String,
    default: 'Close'
  }
}
