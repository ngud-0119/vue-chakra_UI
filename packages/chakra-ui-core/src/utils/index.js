export { default as Logger } from './logger'
export { provideTheme } from './provide-theme'
export { transformAlias as tx, valueToPercent } from './transform'
export { pickProperty as forwardProps, filterBaseStyles, filterPseudo, unwrapValues } from './object'
export * from './color'
export { parsePackIcons } from './icons'
export * from './dom'
export * from './vdom'
export * from './strings'
export * from './generators'
export * from './validators'
export * from './functions'
export * from './components'
export { proxyAliases as pxls } from '../config/props/proxy'
