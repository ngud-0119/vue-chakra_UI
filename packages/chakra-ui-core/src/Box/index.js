
import { css } from 'emotion'
import { background, border, color, borderRadius, flexbox, grid, layout, position, shadow, space, typography, compose } from 'styled-system'
import { baseProps, propsConfig } from '../config/props'
import { forwardProps } from '../utils'

const baseEllipsis = {
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
}

/**
 * @description Truncates text if `truncate` is set to true.
 * @param {Object} props Props
 */
const truncate = props => {
  if (props.truncate) {
    if (!props.lineClamp) {
      return baseEllipsis
    }
  }
}

/**
 * @description Clamps text based on number of lines.
 * @param {Object} props Props
 */
const clamp = props => {
  if (props.lineClamp) {
    return {
      ...baseEllipsis,
      '-webkit-box-orient': 'vertical',
      '-webkit-line-clamp': `${props.lineClamp}`
    }
  }
}

const decorate = props => {
  if (props.textDecoration || props.textDecor) {
    return {
      'text-decoration': `${props.textDecoration || props.textDecor}`
    }
  }
}

const systemProps = compose(
  space,
  layout,
  color,
  background,
  border,
  borderRadius,
  grid,
  position,
  shadow,
  decorate,
  typography,
  flexbox,
  propsConfig,
  truncate,
  clamp
)

const Box = {
  name: 'Box',
  inject: ['$theme'],
  props: {
    as: {
      type: [String, Object],
      default: 'div'
    },
    ...baseProps
  },
  computed: {
    theme () {
      return this.$theme()
    }
  },
  render (h) {
    const { as, ...cleanedStyleProps } = forwardProps(this.$props)
    const boxStylesObject = systemProps({ ...cleanedStyleProps, theme: this.theme })
    const className = css(boxStylesObject)

    return h(as, {
      class: [className]
    }, this.$slots.default)
  }
}

export default Box
