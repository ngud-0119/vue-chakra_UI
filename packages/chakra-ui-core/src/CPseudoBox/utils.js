import map from 'lodash-es/map'
import { forwardProps, filterBaseStyles, filterPseudo as getPseudoStyles, tx } from '../utils'

/**
 * PseudoBox pseudo selectors
 */
export const selectors = {
  _hover: '&:hover',
  _active: '&:active, &[data-active=true]',
  _focus: '&:focus',
  _visited: '&:visited',
  _even: '&:nth-of-type(even)',
  _odd: '&:nth-of-type(odd)',
  _disabled: '&:disabled, &:disabled:focus, &:disabled:hover, &[aria-disabled=true], &[aria-disabled=true]:focus, &[aria-disabled=true]:hover',
  _checked: '&[aria-checked=true]',
  _mixed: '&[aria-checked=mixed]',
  _selected: '&[aria-selected=true]',
  _invalid: '&[aria-invalid=true]',
  _pressed: '&[aria-pressed=true]',
  _readOnly: '&[aria-readonly=true], &[readonly]',
  _first: '&:first-of-type',
  _last: '&:last-of-type',
  _expanded: '&[aria-expanded=true]',
  _grabbed: '&[aria-grabbed=true]',
  _notFirst: '&:not(:first-of-type)',
  _notLast: '&:not(:last-of-type)',
  _groupHover: '[role=group]:hover &',
  _before: '&:before',
  _after: '&:after',
  _focusWithin: '&:focus-within',
  _placeholder: '&::placeholder'
}

/**
 * Filter undefined props and parse pseudo style props to generate css styles object
 * @param {Object} props - Props
 * @returns {Object} Parsed pseudo styles object
 */
export function parsePseudoStyles (props) {
  const pseudoStyles = {}
  const clean = forwardProps(props)
  const pseudoProps = getPseudoStyles(clean)
  const baseProps = filterBaseStyles(clean)
  const result = map(pseudoProps, (value, prop) => ({ prop, value }))
  result.forEach((pair) => {
    if (selectors[pair.prop]) {
      pseudoStyles[selectors[pair.prop]] = tx(pair.value)
    }
  })
  const mergedPseudoBoxStyles = { pseudoStyles, baseProps }
  return mergedPseudoBoxStyles
}
