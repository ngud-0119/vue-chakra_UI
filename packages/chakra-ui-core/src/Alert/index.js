import Box from '../Box'
import Icon from '../Icon'
import { baseProps } from '../config/props'
import { forwardProps } from '../utils'
import useAlertStyle, { useAlertIconStyle } from './alert.styles'

export const statuses = {
  info: { icon: '_info', color: 'blue' },
  warning: { icon: '_warning-2', color: 'orange' },
  success: { icon: '_check-circle', color: 'green' },
  error: { icon: '_warning', color: 'red' }
}

const Alert = {
  name: 'Alert',
  inject: ['$theme', '$colorMode'],
  provide () {
    return {
      _status: this.status,
      _variant: this.variant
    }
  },
  computed: {
    colorMode () {
      return this.$colorMode()
    }
  },
  props: {
    status: {
      type: [String, Array],
      default: 'info'
    },
    variant: {
      type: [String, Array],
      default: 'subtle'
    },
    ...baseProps
  },
  render (h) {
    const alertStyles = useAlertStyle({
      variant: this.variant,
      color: statuses[this.status] && statuses[this.status]['color'],
      colorMode: this.colorMode,
      theme: this.$theme()
    })

    return h(Box, {
      props: {
        fontFamily: 'body',
        ...alertStyles,
        ...forwardProps(this.$props)
      },
      attrs: {
        role: 'alert'
      }
    }, this.$slots.default)
  }
}

const AlertIcon = {
  name: 'AlertIcon',
  inject: ['_status', '_variant', '$colorMode', '$theme'],
  props: {
    size: {
      default: 5
    },
    name: String,
    ...baseProps
  },
  computed: {
    colorMode () {
      return this.$colorMode()
    }
  },
  render (h) {
    const alertIconStyles = useAlertIconStyle({
      variant: this._variant,
      colorMode: this.colorMode,
      color: statuses[this._status] && statuses[this._status]['color']
    })

    return h(Icon, {
      props: {
        mr: this.$props.mr || 3,
        size: this.size,
        name: this.name || (statuses[this._status] && statuses[this._status]['icon']),
        ...alertIconStyles,
        ...forwardProps(this.$props)
      },
      attrs: {
        focusable: false
      }
    })
  }
}

const AlertTitle = {
  name: 'AlertTitle',
  props: {
    ...baseProps
  },
  render (h) {
    return h(Box, {
      props: {
        fontWeight: 'bold',
        lineHeight: 'normal',
        ...forwardProps(this.$props)
      }
    }, this.$slots.default)
  }
}

const AlertDescription = {
  name: 'AlertDescription',
  props: {
    ...baseProps
  },
  render (h) {
    return h(Box, { props: forwardProps(this.$props) }, this.$slots.default)
  }
}

export { Alert, AlertIcon, AlertTitle, AlertDescription }
