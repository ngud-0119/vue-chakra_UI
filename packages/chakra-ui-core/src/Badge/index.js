import Box from '../Box'
import { forwardProps } from '../utils'
import { baseProps } from '../config/props'
import useBadgeStyles from './badge.styles'

export default {
  name: 'Badge',
  inject: ['$theme', '$colorMode'],
  props: {
    variant: {
      type: String,
      default: 'subtle'
    },
    variantColor: {
      type: String,
      default: 'gray'
    },
    ...baseProps
  },
  computed: {
    colorMode () {
      return this.$colorMode()
    }
  },
  render (h) {
    const badgeStyleProps = useBadgeStyles({
      theme: this.$theme(),
      colorMode: this.colorMode,
      color: this.variantColor,
      variant: this.variant
    })

    return h(Box, {
      props: {
        ...forwardProps(this.$props),
        d: 'inline-block',
        textTransform: 'uppercase',
        fontSize: 'xs',
        fontFamily: 'body',
        px: 1,
        rounded: 'sm',
        fontWeight: 'bold',
        whiteSpace: 'nowrap',
        verticalAlign: 'middle',
        ...badgeStyleProps
      }
    }, this.$slots.default)
  }
}
