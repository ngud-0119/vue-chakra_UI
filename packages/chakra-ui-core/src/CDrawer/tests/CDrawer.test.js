import Vue from 'vue'
import { CInput, CButton, CDrawer, CDrawerBody, CDrawerFooter, CDrawerHeader, CDrawerOverlay, CDrawerContent, CDrawerCloseButton } from '../../'
import { render, userEvent, fireEvent, waitMs } from '@/tests/test-utils'
import { useId } from '@/packages/chakra-ui-core/src/utils'

// mocks
jest.mock('@/packages/chakra-ui-core/src/utils/generators.js')

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { CInput, CButton, CDrawer, CDrawerBody, CDrawerFooter, CDrawerHeader, CDrawerOverlay, CDrawerContent, CDrawerCloseButton },
    data: () => ({
      isOpen: false
    }),
    methods: {
      close () {
        this.isOpen = false
      }
    },
    template: `
    <div>
      <CInput data-testid="inputOutsideDrawer" ref="inputOutsideDrawer" placeholder="Type here..." /> 
      <CButton data-testid="buttonOutside" ref="btnRef" @click="isOpen = true">Open Drawer</CButton>

      <CDrawer placement="right" ${inlineAttrs}>
        <CDrawerOverlay />
        <CDrawerContent data-testid="overlay">
          <CDrawerCloseButton data-testid="close-btn" />
          <CDrawerHeader>Create your account</CDrawerHeader>
          <CDrawerBody>
            <CInput data-testid="inputInsideDrawer" ref="inputInsideDrawer" placeholder="Type here..." />
          </CDrawerBody>
          <CDrawerFooter>
            <CButton variant="outline" mr="3" @click="isOpen = false">Cancel</CButton>
            <CButton color="blue">Save</CButton>
          </CDrawerFooter>
        </CDrawerContent>
      </CDrawer>
    </div>
    `,
    ...props
  }
  return render(base, {}, vue => vue.directive('scroll-lock', jest.fn()))
}

it('should render correctly', async () => {
  useId.mockReturnValueOnce('1')
  const inlineAttrs = 'isOpen'
  const { asFragment } = renderComponent({ inlineAttrs })

  await Vue.nextTick()

  expect(asFragment(document.body.innerHTML)).toMatchSnapshot()
})

test('clicking the close button calls the onClose callback', async () => {
  const onClose = jest.fn()
  const inlineAttrs = 'isOpen :on-close="close"'
  const { getByTestId } = renderComponent({ inlineAttrs, methods: { close: onClose } })

  await Vue.nextTick()
  userEvent.click(getByTestId('close-btn'))

  expect(onClose).toHaveBeenCalled()
})

test('pressing "esc" calls the onClose callback', async () => {
  const onClose = jest.fn()
  const inlineAttrs = ':isOpen="isOpen" :on-close="close"'
  const { getByTestId } = renderComponent({ inlineAttrs, data: () => ({ isOpen: true }), methods: { close: onClose } })

  await Vue.nextTick()
  const inputInside = getByTestId('inputInsideDrawer')

  fireEvent.keyDown(inputInside, { key: 'Escape' })

  expect(onClose).toHaveBeenCalled()
})

test('clicking overlay calls the onClose callback', async () => {
  const onClose = jest.fn()
  const inlineAttrs = ':isOpen="isOpen" :on-close="close"'
  const { getByTestId } = renderComponent({ inlineAttrs, data: () => ({ isOpen: true }), methods: { close: onClose } })

  await Vue.nextTick()
  const overlay = getByTestId('overlay')

  userEvent.click(overlay)

  expect(onClose).toHaveBeenCalled()
})

test('focuses the initial focus ref when opened - initialFocusRef', async () => {
  const inlineAttrs = 'isOpen :on-close="close" :initialFocusRef="()=>$refs.inputInsideDrawer"'
  const { getByTestId } = renderComponent({ inlineAttrs })

  await Vue.nextTick()
  const inputInside = getByTestId('inputInsideDrawer')

  await waitMs()
  expect(inputInside).toHaveFocus()
})

test('returns focus when closed - finalFocusRef', async () => {
  const inlineAttrs = ':isOpen="isOpen" :on-close="close" :finalFocusRef="$refs.inputOutsideDrawer"'
  const { getByTestId } = renderComponent({ inlineAttrs, data: () => ({ isOpen: true }) })

  await Vue.nextTick()
  const inputOutside = getByTestId('inputOutsideDrawer')
  const closeButton = getByTestId('close-btn')

  await userEvent.click(closeButton)
  await waitMs()

  expect(inputOutside).toHaveFocus()
})

it('should have proper aria', async () => {
  useId.mockReturnValueOnce('1')
  const inlineAttrs = 'isOpen'
  renderComponent({ inlineAttrs })

  await Vue.nextTick()
  const dialog = document.querySelector('section')
  await waitMs()

  expect(dialog).toHaveAttribute('role', 'dialog')
  expect(dialog).toHaveAttribute('aria-modal', 'true')
  expect(dialog).toHaveAttribute('aria-labelledby', 'drawer-1-header')
  expect(dialog).toHaveAttribute('aria-describedby', 'drawer-1-body')
})
