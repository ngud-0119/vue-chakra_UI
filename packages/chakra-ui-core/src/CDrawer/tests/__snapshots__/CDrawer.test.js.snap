// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        class="css-0 css-1l17jza"
        data-chakra-component="CInput"
        data-testid="inputOutsideDrawer"
        placeholder="Type here..."
      />
       
      <button
        class="css-0 css-rup5mm"
        data-chakra-component="CButton"
        data-testid="buttonOutside"
        tabindex="0"
        type="button"
      >
        Open Drawer
      </button>
       
      <!---->
    </div>
  </div>
  <span
    id="modal-portal-1"
  >
    <div
      class="css-0 css-79elbk"
      data-chakra-component="CPseudoBox"
    >
      <div
        style="position: fixed; top: 0px; right: 0px; bottom: 0px; left: 0px;"
      >
        <div
          class="css-l5w8ps"
          data-chakra-component="CDrawerOverlay"
        />
        <div
          class="css-13c8a1c"
          data-chakra-component="CModalContent"
          data-testid="overlay"
        >
          <section
            aria-describedby="drawer-1-body"
            aria-labelledby="drawer-1-header"
            aria-modal="true"
            class="css-iv9bja"
            data-chakra-component="CDrawerContent"
            id="drawer-1"
            role="dialog"
            tabindex="-1"
          >
            <button
              aria-label="Close"
              class="css-0 css-1xo4iay"
              data-chakra-component="CDrawerCloseButton"
              data-testid="close-btn"
              x-close-button=""
            >
              <svg
                aria-hidden="true"
                class="css-xmst1h css-y3asau"
                data-chakra-component="CIcon"
                role="presentation"
                viewBox="0 0 24 24"
              >
                
      
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
             
            <header
              class="css-kniazf"
              data-chakra-component="CModalHeader"
              id="drawer-1-header"
            >
              Create your account
            </header>
             
            <div
              class="css-5m10fw"
              data-chakra-component="CModalBody"
              id="drawer-1-body"
            >
              <input
                class="css-0 css-1l17jza"
                data-chakra-component="CInput"
                data-testid="inputInsideDrawer"
                placeholder="Type here..."
              />
            </div>
             
            <footer
              class="css-plnu2w"
              data-chakra-component="CModalFooter"
            >
              <button
                class="css-0 css-1tex3qd"
                data-chakra-component="CButton"
                tabindex="0"
                type="button"
              >
                Cancel
              </button>
               
              <button
                class="css-0 css-2jmp8f"
                data-chakra-component="CButton"
                tabindex="0"
                type="button"
              >
                Save
              </button>
            </footer>
          </section>
        </div>
      </div>
    </div>
  </span>
</DocumentFragment>
`;
